### DISCUSSION TITLE
Connections warming up

### DISCUSSION THREAD

**Discussion by Brieux1** [netty/netty#13841] [Q&A] [answered]
Hello,
We are migrating an application to Spring Webflux and are having questions concerning the netty client warm up.
Our application needs a lot of time/requests before the netty is ready.

Context:
- our application exposes 1 web service through a rest controller
- the application then make several call to an other application (simulated by a wiremock application in our testing environment)

Our testing environment: 
- kubernetes hosted on AWS
- application: Spring Webflux (jdk 17) 1 pod
- mocked partner: 5 pods in the same namespace

Our webclient configuration:
`ConnectionProvider provider = ConnectionProvider
        .builder("custom-http")
        .pendingAcquireTimeout(Duration.ofSeconds(10))
        .maxConnections(200)
        .pendingAcquireMaxCount(10000)
        .maxIdleTime(Duration.ofSeconds(30))
        .maxLifeTime(Duration.ofSeconds(60))
        .evictInBackground(Duration.ofSeconds(120))
        .disposeInactivePoolsInBackground(Duration.ofSeconds(180), Duration.ofSeconds(175))
        .build();

    HttpClient httpClient = HttpClient.create(provider)
        .metrics(true, s -> s)
        .resolver(DefaultAddressResolverGroup.INSTANCE)
        .option(ChannelOption.CONNECT_TIMEOUT_MILLIS, 2000)
        .responseTimeout(Duration.ofMillis(2000))
        .doOnConnected(conn -> {
              conn.addHandlerLast(new IdleStateHandler(3000, 3000, 3000, TimeUnit.MILLISECONDS));
              conn.addHandlerLast(new WriteTimeoutHandler(5000, TimeUnit.MILLISECONDS));
              conn.addHandlerLast(new ReadTimeoutHandler(5000, TimeUnit.MILLISECONDS));
            }
        );`

We warm up the application by calling the web service manually several times (~10 times) and the responses are OK.

Then we start a load testrun (~400 requests/sec for 1 minute).
The results are KO (see failed_run screenshot attached).
![failed_run](https://github.com/netty/netty/assets/147488411/2ae0ed06-a0b2-42bc-bc9c-6c1e4b660a06)

Then we launch the same test after a few seconds and the results are OK (see successful_run screenshot attached).
![successful_run](https://github.com/netty/netty/assets/147488411/98513fc0-b5b7-4afb-a29f-f08f30ab6764)

Here are the measured we retrieved during the 2 testruns :
- first run: from 09:11:45 to 09:13:10
- second run: from 09:14:54 to 09:16:15

We can see that on the first run, the netty client encounters a lot of "client error" (see client_error screenshot).
![client_error](https://github.com/netty/netty/assets/147488411/42ebcbef-fe63-4424-9d6c-37f35f2720f6)

And that the netty client use nearly 200 active connections on the first run, and stacking requests in pending connections.
Whereas on the second run only 45 are necessary and we retrieve 0 pending connections.

Why does the client needs so much time/requests to be ready?
We tested several configuration (including the warmup() method on the client) without improvements.

**Comment by slandelle** [comment, marked as answer]
Hello,

I think this is a question for the Spring Webflux as it’s about their HTTP client.

**Reply by Brieux1** [reply]
Thank you for the response.
We posted our question in their github.
